webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/volume/VolumeTracingDownsampling.scala

```diff
@@ -0,0 +1,209 @@
+        sourceBucketPositionsFor(downsampledBucketPosition, downScaleFactor, previousMag)
+      val sourceData: Seq[Array[Byte]] = sourceBuckets.map(bucketDataMap(_))
+      val downsampledData: Array[Byte] =
+        if (sourceData.forall(_.sameElements(Array[Byte](0))))
+          Array[Byte](0)
+        else {
+          val sourceDataFilled = fillZeroedIfNeeded(sourceData, bucketVolume, dataLayer.bytesPerElement)
+          val sourceDataTyped = UnsignedIntegerArray.fromByteArray(sourceDataFilled.toArray.flatten, elementClass)
+          val dataDownscaledTyped =
+            downsampleData(sourceDataTyped.grouped(bucketVolume).toArray, downScaleFactor, bucketVolume)
+          UnsignedIntegerArray.toByteArray(dataDownscaledTyped, elementClass)
+        }
+      bucketDataMap(downsampledBucketPosition) = downsampledData
+      updatedBuckets.add(downsampledBucketPosition)
+    }
+  }
+
+  private def downsampledBucketPositions(originalBucketPositions: List[BucketPosition],
+                                         requiredMag: Point3D): Set[BucketPosition] =
+    originalBucketPositions.map { bucketPosition: BucketPosition =>
+      BucketPosition(
+        (bucketPosition.globalX / requiredMag.x / 32) * requiredMag.x * 32,
+        (bucketPosition.globalY / requiredMag.y / 32) * requiredMag.y * 32,
+        (bucketPosition.globalZ / requiredMag.z / 32) * requiredMag.z * 32,
+        requiredMag
+      )
+    }.toSet
+
+  private def sourceBucketPositionsFor(bucketPosition: BucketPosition,
+                                       downScaleFactor: Point3D,
+                                       previousMag: Point3D): Seq[BucketPosition] =
+    for {
+      z <- 0 until downScaleFactor.z
+      y <- 0 until downScaleFactor.y
+      x <- 0 until downScaleFactor.x
+    } yield {
+      BucketPosition(
+        bucketPosition.globalX + x * bucketPosition.bucketLength * previousMag.x,
+        bucketPosition.globalY + y * bucketPosition.bucketLength * previousMag.y,
+        bucketPosition.globalZ + z * bucketPosition.bucketLength * previousMag.z,
+        previousMag
+      )
+    }
+
+  private def fillZeroedIfNeeded(sourceData: Seq[Array[Byte]],
+                                 bucketVolume: Int,
+                                 bytesPerElement: Int): Seq[Array[Byte]] =
+    // Reverted buckets and missing buckets are represented by a single zero-byte.
+    // For downsampling, those need to be replaced with the full bucket volume of zero-bytes.
+    sourceData.map { sourceBucketData =>
+      if (sourceBucketData.sameElements(Array[Byte](0))) {
+        Array.fill[Byte](bucketVolume * bytesPerElement)(0)
+      } else sourceBucketData
+    }
+
+  private def downsampleData[T: ClassTag](data: Array[Array[T]],
+                                          downScaleFactor: Point3D,
+                                          bucketVolume: Int): Array[T] = {
+    val result = new Array[T](bucketVolume)
+    for {
+      z <- 0 until 32
```

Comment:
Extract magic number

Reply:
@youri-k Was this a note for yourself? Or for @fm3 ?

Reply:
It was for @fm3, but we decided in person to leave it as it is.